File: backend/test/shared/validation.test.ts
import { describe, it, expect } from 'vitest';
import {
  isHoneypotTripped,
  isTooFast,
  isValidEmail,
  isValidMessage,
  sanitizeName,
  MIN_FORM_TIME_MS,
  MAX_MESSAGE,
} from '../../src/shared/validation';

describe('isHoneypotTripped', () => {
  it('passes when website is empty', () => {
    expect(isHoneypotTripped({ website: '' })).toBe(false);
    expect(isHoneypotTripped({ website: '   ' })).toBe(false);
  });
  it('trips when website is filled', () => {
    expect(isHoneypotTripped({ website: 'http://spam' })).toBe(true);
  });
});

describe('isTooFast', () => {
  it('rejects submissions faster than the minimum', () => {
    const now = 1_000_000;
    expect(isTooFast(now - (MIN_FORM_TIME_MS - 1), now)).toBe(true);
  });
  it('allows submissions at or past the minimum', () => {
    const now = 1_000_000;
    expect(isTooFast(now - MIN_FORM_TIME_MS, now)).toBe(false);
  });
});

describe('isValidEmail', () => {
  it('accepts a normal address', () => {
    expect(isValidEmail('a@b.co')).toBe(true);
  });
  it('rejects malformed or oversized addresses', () => {
    expect(isValidEmail('nope')).toBe(false);
    expect(isValidEmail('a@b')).toBe(false);
    expect(isValidEmail('x'.repeat(255) + '@b.co')).toBe(false);
  });
});

describe('isValidMessage', () => {
  it('accepts a valid non-empty string', () => {
    expect(isValidMessage('hello there')).toBe(true);
  });
  it('rejects an empty string', () => {
    expect(isValidMessage('')).toBe(false);
  });
  it('rejects a non-string (number)', () => {
    expect(isValidMessage(123)).toBe(false);
  });
  it('rejects a string over MAX_MESSAGE length', () => {
    expect(isValidMessage('a'.repeat(MAX_MESSAGE + 1))).toBe(false);
  });
  it('accepts a string exactly at MAX_MESSAGE length', () => {
    expect(isValidMessage('a'.repeat(MAX_MESSAGE))).toBe(true);
  });
});

describe('sanitizeName', () => {
  it('strips CR, LF, tabs and control chars (header-injection fix)', () => {
    expect(sanitizeName('Alice\nBcc: evil@x.com')).toBe('AliceBcc: evil@x.com');
    expect(sanitizeName('A\r\nB\tC')).toBe('ABC');
  });
  it('caps length at 200 chars', () => {
    expect(sanitizeName('a'.repeat(500)).length).toBe(200);
  });
});

File: backend/internal/validation/validation_test.go
package validation_test

import (
	"strings"
	"testing"

	"github.com/Grimm07/portfolio-backend/backend/internal/validation"
)

func TestIsHoneypotTripped(t *testing.T) {
	tests := []struct {
		name    string
		website string
		want    bool
	}{
		{"empty string", "", false},
		{"whitespace only", "   ", false},
		{"filled", "http://spam", true},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if got := validation.IsHoneypotTripped(tc.website); got != tc.want {
				t.Errorf("IsHoneypotTripped(%q) = %v, want %v", tc.website, got, tc.want)
			}
		})
	}
}

func TestIsTooFast(t *testing.T) {
	const now = int64(1_000_000)
	tests := []struct {
		name          string
		formTimestamp int64
		want          bool
	}{
		{
			name:          "just under minimum — rejected",
			formTimestamp: now - (validation.MinFormTimeMs - 1),
			want:          true,
		},
		{
			name:          "exactly at minimum — allowed",
			formTimestamp: now - validation.MinFormTimeMs,
			want:          false,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if got := validation.IsTooFast(tc.formTimestamp, now); got != tc.want {
				t.Errorf("IsTooFast(%d, %d) = %v, want %v", tc.formTimestamp, now, got, tc.want)
			}
		})
	}
}

func TestIsValidEmail(t *testing.T) {
	oversized := strings.Repeat("x", 255) + "@b.co"
	tests := []struct {
		name  string
		email string
		want  bool
	}{
		{"normal address", "a@b.co", true},
		{"no at-sign", "nope", false},
		{"missing TLD", "a@b", false},
		{"oversized local part", oversized, false},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if got := validation.IsValidEmail(tc.email); got != tc.want {
				t.Errorf("IsValidEmail(%q) = %v, want %v", tc.email, got, tc.want)
			}
		})
	}
}

func TestIsValidMessage(t *testing.T) {
	tests := []struct {
		name    string
		message string
		want    bool
	}{
		{"valid non-empty", "hello there", true},
		{"empty string", "", false},
		{"over MaxMessage", strings.Repeat("a", validation.MaxMessage+1), false},
		{"exactly MaxMessage", strings.Repeat("a", validation.MaxMessage), true},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if got := validation.IsValidMessage(tc.message); got != tc.want {
				t.Errorf("IsValidMessage(%q...) = %v, want %v", tc.message[:min(len(tc.message), 20)], got, tc.want)
			}
		})
	}
}

func TestSanitizeName(t *testing.T) {
	t.Run("strips CR LF tab and control chars", func(t *testing.T) {
		cases := []struct{ in, want string }{
			{"Alice\nBcc: evil@x.com", "AliceBcc: evil@x.com"},
			{"A\r\nB\tC", "ABC"},
		}
		for _, c := range cases {
			if got := validation.SanitizeName(c.in); got != c.want {
				t.Errorf("SanitizeName(%q) = %q, want %q", c.in, got, c.want)
			}
		}
	})
	t.Run("caps at MaxName runes", func(t *testing.T) {
		result := validation.SanitizeName(strings.Repeat("a", 500))
		runes := []rune(result)
		if len(runes) != validation.MaxName {
			t.Errorf("SanitizeName(500×a) rune length = %d, want %d", len(runes), validation.MaxName)
		}
	})
}

// min is a local helper for older Go toolchains (pre-1.21 builtin).
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
